ntbs-service/Services/AdImportService.cs

```diff
@@ -0,0 +1,82 @@
+
+        public AdImportService(IAdDirectoryFactory adDirectoryFactory, IReferenceDataRepository referenceDataRepository, IUserRepository userRepository)
+        {
+            _adDirectoryFactory = adDirectoryFactory;
+            _referenceDataRepository = referenceDataRepository;
+            _userRepository = userRepository;
+        }
+
+        // get those values from settings
+        public async Task RunCaseManagerImport()
+        {
+            var tbServices = await _referenceDataRepository.GetAllTbServicesAsync();
+            using (var adDirectoryService = _adDirectoryFactory.Create())
+            {
+                foreach (var directoryEntry in adDirectoryService.GetAllDirectoryEntries())
+                {
+                    var userName = adDirectoryService.GetUsername(directoryEntry);
+                    if (userName == null)
+                    {
+                        continue;
+                    }
+
+                    var userPrincipal = adDirectoryService.GetUserPrincipal(userName);
+                    var user = new User
+                    {
+                        Username = userPrincipal.UserPrincipalName,
+                        GivenName = userPrincipal.GivenName,
+                        FamilyName = userPrincipal.Surname,
+                        DisplayName = userPrincipal.DisplayName,
+                        IsActive = adDirectoryService.IsUserEnabled(directoryEntry)
+                    };
+
+                    user.SetAdGroups(GetAdGroups(adDirectoryService, directoryEntry));
+                    var filteredTbServices = tbServices.Where(tb => user.AdGroupNames.Contains(tb.ServiceAdGroup));
+                    user.IsCaseManager = filteredTbServices.Any();
+
+                    await _userRepository.AddOrUpdateUser(user, filteredTbServices);
+                }  
+            }
+        }
+
+        private static List<string> GetAdGroups(AdDirectoryService adDirectoryService, DirectoryEntry directoryEntry)
+        {
+            var adGroups = new List<string>();
+            // The more natural way to obtain membership groups would have been to use user.GetAuthorizationGroups()
+            // However, all operations trying to fetch AD groups failed with an error "PrincipalOperationException: Information about the domain could not be retrieved"
+            // This is caused by not being in the same network as the AD server. None of the suggested workarounds worked, so fetching this via this property instead
+            foreach (var distinguishedName in adDirectoryService.GetDistinguisedGroupNames(directoryEntry))
+            {
+                var match = DistinguishedNameRegex.Match(distinguishedName);
+                if (!match.Success)
+                {
+                    continue;
+                }
+                adGroups.Add(match.Groups[2].Captures[0].Value);
+            }
+            return adGroups;
+        }

```

Comment:
So what are the `distinguishedName` values? why can't we just use those directly?

Reply:
They look a bit like uid=john.doe,ou=People,dc=example,dc=com (can't remember the exact syntax).

Reply:
ah right, I was reading the regex wrong (as if it was capturing the entire string)!